halo2-wasm/src/halo2lib.rs

```diff
@@ -328,6 +329,77 @@ impl Halo2LibWasm {
         self.to_js_assigned_values(out)
     }
 
+    pub fn to_hi_lo(&mut self, a: usize) -> Vec<u32> {
+        let a = self.get_assigned_value(a);
+        let a_val = a.value();
+        let a_bytes = a_val.to_bytes();
+
+        let mut a_lo_bytes = [0u8; 32];
+        let mut a_hi_bytes = [0u8; 32];
+        a_lo_bytes[..16].copy_from_slice(&a_bytes[..16]);
+        a_hi_bytes[..16].copy_from_slice(&a_bytes[16..]);
+        let a_lo = Fr::from_bytes(&a_lo_bytes).unwrap();
+        let a_hi = Fr::from_bytes(&a_hi_bytes).unwrap();
+
+        let a_lo = self.builder.borrow_mut().main(0).load_witness(a_lo);
+        let a_hi = self.builder.borrow_mut().main(0).load_witness(a_hi);
+
+        let two_pow_128 = self.gate.pow_of_two()[128];
+        let r = modulus::<Fr>();
+        let a_hi_max = r.clone() / fe_to_biguint(&two_pow_128);
+        let a_lo_max = r % fe_to_biguint(&two_pow_128);
+
+        //check a_hi < r // 2**128
+        let check_1 = self.range.is_big_less_than_safe(
+            self.builder.borrow_mut().main(0),
+            a_hi,
+            a_hi_max.clone(),
+        );
+
+        //check (a_hi == r // 2 ** 128 AND a_lo < r % 2**128)
+        let a_hi_max_fe = biguint_to_fe::<Fr>(&a_hi_max);
+        let check_2_hi = self.gate.is_equal(
+            self.builder.borrow_mut().main(0),
+            a_hi,
+            Constant(a_hi_max_fe),
+        );
+        let check_2_lo =
+            self.range
+                .is_big_less_than_safe(self.builder.borrow_mut().main(0), a_lo, a_lo_max);
+        let check_2 = self
+            .gate
+            .and(self.builder.borrow_mut().main(0), check_2_hi, check_2_lo);
+
+        //constrain (check_1 || check_2) == 1
+        let check = self
+            .gate
+            .or(self.builder.borrow_mut().main(0), check_1, check_2);
```

Comment:
since `check_1` and `check_2` are mutually exclusive, you can just do `check_1 + check_2`